### PR TITLE
[CGP-276] Typechecker: handle redexes in all places

### PR DESCRIPTION
**Pull request by michaelpj**
- We were reporting the wrong type in an error (test on this commit gives wrong message before that, cf CGP-268)
- Type substitution should cost us.
- Reduce types at all times before checking rule applicability.
- Reduce iterated applications correctly.
    - Currently we will reduce one application, but then we will think that we're done, when in fact we have to do it repeatedly until the rule no longer applies.

I *think* this is correct, and it handles the cases I'm interested in at least. This does *not* ensure that the type inferred for a term is reduced, I'm unsure whether that's desirable.

### CODE REVIEW

**Comment by michaelpj** [comment]
- Remove the reduction after substitution for wrap/unwrap, we should reduce when we need it.

Actually, I'm pretty sure this is still lacking, since we should reduce both sides every time we check for type equality.

**Comment by michaelpj** [comment]
Okay, I think this is now reasonable, but I'm going to write a bunch more tests.

**Comment by mchakravarty** [comment]
> This does not ensure that the type inferred for a term is reduced, I'm unsure whether that's desirable.

Yes, this is not just desirable, but a requirement. It is required by the type synthesis judgement form.

**Comment by michaelpj** [comment]
Yes, the idea is to update the spec too. We certainly shouldn't merge this until we have the corresponding spec update as well.

**Comment by vmchale** [comment]
We should write a function that normalizes types or doesn't based on configuration instead, and merge that  in the meantime.

**Comment by michaelpj** [comment]
The reason we decided to do it this way is that:
- It's hard to implement a type-reducer without kind-checking your program, at which point you might as well type-check it.
- Having the underlying type-checker support this gives us flexibility and brings us closer to baseline System Fw. All we have to do to get the old behaviour is turn on the normalization check.

The recent emails on the list should give some more context for this.

File: language-plutus-core/src/Language/PlutusCore/TypeSynthesis.hs [hunk 1]
```diff
@@ -238,18 +246,26 @@ extractUnique = nameUnique . unTyName . unTyNameWithKind
 tySubstitute :: Unique -- ^ Unique associated with type variable
              -> Type TyNameWithKind a -- ^ Type we are binding to free variable
              -> Type TyNameWithKind a -- ^ Type we are substituting in
-             -> Type TyNameWithKind a
-tySubstitute u ty = cata a where
+             -> TypeCheckM b (Type TyNameWithKind a)
+tySubstitute u ty t = typeCheckStep >> (pure $ cata a t) where
```
Comment (vmchale) [inline language-plutus-core/src/Language/PlutusCore/TypeSynthesis.hs]: We should probably reenable the `hlint` for "move parentheses to avoid $" or whatever, since this could be

```
typeCheckStep >> pure (cata a t)
```

**Comment by vmchale** [comment]
The bug report says we should have two modes for the type checker, no?

**Comment by michaelpj** [comment]
I don't think that's correct, I'll comment there.

File: language-plutus-core/src/Language/PlutusCore/Renamer.hs [hunk 2]
```diff
@@ -80,9 +80,9 @@ annotateT (Constant x c) =
 annotateT (TyInst x t ty) =
     TyInst x <$> annotateT t <*> annotateTy ty
 annotateT (Wrap x (TyName (Name x' b u@(Unique i))) ty t) = do
-    aty <- annotateTy ty
     let k = Type x'
     insertKind i k
+    aty <- annotateTy ty
```
Comment (vmchale) [inline language-plutus-core/src/Language/PlutusCore/Renamer.hs]: What does this do?

Comment (michaelpj) [reply language-plutus-core/src/Language/PlutusCore/Renamer.hs]: Read the commit message :) this is an unrelated bug which we didn't notice because there are no tests for wrap/unwrap. I'm going to pull them out in a bit.

File: language-plutus-core/test/types/recursiveWrap.plc.golden [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+        f
+        (fun a (fun (fix list (all r (type) (fun r (fun (fun a (fun list r)) r)))) r))
+        z
+      )
+    )
+  )'.
+Expected type
+  '(all r (type) (fun r (fun (fun a (fun (fix list (all r (type) (fun r (fun (fun a (fun list r)) r)))) r)) r)))',
+found type
+  '(all r (type) (fun r (fun (fun a (fun (fix list (all r (type) (fun r (fun (fun a (fun list r)) r)))) r)) r)))'
```
Comment (vmchale) [inline language-plutus-core/test/types/recursiveWrap.plc.golden]: It looks like type equality or pretty-printing needs to be fixed. Should I do this in a separate PR?

Comment (michaelpj) [reply language-plutus-core/test/types/recursiveWrap.plc.golden]: I've headed home for the evening but yeah, this is where I'd got stuck - it does look rather suspicious!

Comment (vmchale) [reply language-plutus-core/test/types/recursiveWrap.plc.golden]: Sounds good. I will try to fix this. I'm not sure if pretty-printing ran awry or equality :\

Comment (michaelpj) [reply language-plutus-core/test/types/recursiveWrap.plc.golden]: You'll need the renamer fix too.

**Comment by michaelpj** [comment]
This needs work, so closing for now.
